lib/vtls/rustls.c

```diff
@@ -0,0 +1,409 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2020, Jacob Hoffman-Andrews, <github@hoffman-andrews.com>
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+#include "curl_setup.h"
+
+#ifdef USE_RUSTLS
+
+#include "curl_printf.h"
+
+#include <errno.h>
+#include <crustls.h>
+
+#include "urldata.h"
+#include "sendf.h"
+#include "vtls.h"
+
+struct ssl_backend_data
+{
+  struct rustls_client_session *session;
+  bool data_pending;
+};
+
+static const struct rustls_client_config *client_config = NULL;
+
+static int
+Curl_rustls_init(void)
```

Comment:
Just noticed! We use the `Curl_` prefix for symbols that are global within the library, and since this is static it should **not** use `Curl_` (it goes for a few other functions too). Most TLS backends just go with a fixed private prefix. I would probably not recommend using `rustls_` since that's the prefix the actual TLS library use.

Reply:
It seems to me that local style is to use a `Curl_<backend>_` prefix for the functions that wind up as fields of the `Curl_ssl` struct:

```
$ grep Curl_.*version, lib/vtls/*.c 
lib/vtls/bearssl.c:  Curl_bearssl_version,
lib/vtls/gskit.c:  Curl_gskit_version,             /* version */
lib/vtls/gtls.c:  Curl_gtls_version,             /* version */
lib/vtls/mbedtls.c:  Curl_mbedtls_version,             /* version */
lib/vtls/mesalink.c:  Curl_mesalink_version, /* version */
lib/vtls/nss.c:  Curl_nss_version,             /* version */
lib/vtls/openssl.c:  Curl_ossl_version,             /* version */
lib/vtls/schannel.c:  Curl_schannel_version,             /* version */
lib/vtls/sectransp.c:  Curl_sectransp_version,             /* version */
lib/vtls/vtls.c:  Curl_multissl_version,             /* version */
lib/vtls/wolfssl.c:  Curl_wolfssl_version,             /* version */
```

(each of those functions is static to the relevant file). Am I missing something here?

Reply:
See #6443 ...